### PR TITLE
Redesign of the nurse command

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
Please review each commit by review also the comments of the module.

### CODE REVIEW

File: coffee_core/src/nurse/strategy.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+//! > In this way, a client using the algorithm may choose
+//! > a specific implementation, while the general algorithm
+//! > workflow remains the same. In other words, the abstract
+//! > specification of the class does not depend on the specific
+//! > implementation of the derived class, but specific implementation
+//! > must adhere to the abstract specification.
+//!
+//! So in this specific case the nurse command may need
+//! different kind of recovery algorithm, so we should
+//! be able to choose the algorithm at runtime.
```
Comment (tareknaser360) [inline coffee_core/src/nurse/strategy.rs:21]: Thanks for the explanation

File: coffee_core/src/nurse/chain.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
+//! Nurse Chain of Reponsability rust implementation
```
Comment (tareknaser360) [inline coffee_core/src/nurse/chain.rs:1]:
```suggestion
//! Nurse Chain of Responsibility rust implementation
```

File: coffee_core/src/nurse/chain.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+//!
+//! If you do not know what Chain Of Reposability patter
```
Comment (tareknaser360) [inline coffee_core/src/nurse/chain.rs:3]:
```suggestion
//! If you do not know what Chain Of Repeatability patter
```

File: coffee_core/src/nurse/chain.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+//! proud of me for the following design, or simply mad with me!
+//!
+//! Author: Vincenzo Palazzo <vincenzopalazzo@member.fsf.org>
+use std::sync::Arc;
+
+use async_trait::async_trait;
+
+use coffee_lib::errors::CoffeeError;
+
+use super::strategy::RecoveryStragety;
```
Comment (tareknaser360) [inline coffee_core/src/nurse/chain.rs:35]:
```suggestion
use super::strategy::RecoveryStrategy;
```

File: coffee_core/src/nurse/chain.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+#[async_trait]
+pub trait Handler: Send + Sync {
+    async fn can_be_apply(
+        self: Arc<Self>,
+    ) -> Result<Option<Arc<dyn RecoveryStragety>>, CoffeeError>;
```
Comment (tareknaser360) [inline coffee_core/src/nurse/chain.rs:41]:
```suggestion
    ) -> Result<Option<Arc<dyn RecoveryStrategy>>, CoffeeError>;
```

File: coffee_core/src/nurse/chain.rs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+pub struct RecoveryChainOfResponsability {
```
Comment (tareknaser360) [inline coffee_core/src/nurse/chain.rs:44]:
```suggestion
pub struct RecoveryChainOfResponsibility {
```

File: coffee_core/src/nurse/chain.rs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    pub handlers: Vec<Arc<dyn Handler>>,
+}
+
+impl RecoveryChainOfResponsability {
```
Comment (tareknaser360) [inline coffee_core/src/nurse/chain.rs:48]:
```suggestion
impl RecoveryChainOfResponsibility {
```

File: coffee_core/src/nurse/strategy.rs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+use std::sync::Arc;
+
+use async_trait::async_trait;
+
+use coffee_lib::errors::CoffeeError;
+
+use crate::nurse::chain::Handler;
+
+#[async_trait]
+pub trait RecoveryStragety: Send + Sync {
```
Comment (tareknaser360) [inline coffee_core/src/nurse/strategy.rs:33]:
```suggestion
pub trait RecoveryStrategy: Send + Sync {
```

File: coffee_core/src/nurse/strategy.rs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    async fn patch(&self) -> Result<(), CoffeeError>;
+}
+
+pub struct GitRepositoryMissedStragegy;
```
Comment (tareknaser360) [inline coffee_core/src/nurse/strategy.rs:37]:
```suggestion
pub struct GitRepositoryMissedStrategy;
```

**Review by tareknaser360 — COMMENTED** [review]
These are just typos
The skeleton is very nice
I can work on this as well

**Comment by vincenzopalazzo** [comment]
>I can work on this as well

Yeah I will leave this to you now, let me fix the review and we merge it as a draft

**Comment by vincenzopalazzo** [comment]
Merging this and leaving it to @tareknaser360

**Pull request merged**